include/swift/SILOptimizer/Analysis/ARCAnalysis.h

```diff
@@ -386,4 +390,179 @@ SILInstruction *findReleaseToMatchUnsafeGuaranteedValue(
 
 } // end namespace swift
 
+
+namespace swift {
+
+/// EpilogueARCBlockState - Keep track of whether a epilogue ARC instruction has
+/// been found.
+struct EpilogueARCBlockState {
+  /// Keep track of whether a eplogue release has been found before and after
+  /// this basic block.
+  bool BBSetIn;
+  bool BBSetOut;
+  /// Constructor, we only compute epilogue ARC instruction for 1 argument at
+  /// a time.
+  EpilogueARCBlockState() {
+    // Optimistic data flow.
+    BBSetIn = true;
+    BBSetOut = false;
+  }
+};
+
+/// EpilogueARCContext - This class implements a data flow with which epilogue
+/// retains or releases for a SILValue are found.
+///
+/// NOTE:
+/// In case of release finder, this function assumes the SILArgument has
+/// @owned semantic.
+/// In case of retain finder, this class assumes Arg is one of the return value
+/// of the function.
+class EpilogueARCContext {
+public:
+  enum EpilogueARCKind { Retain = 0, Release = 1 };
+
+private:
+  // Are we finding retains or releases.
+  EpilogueARCKind Kind;
+
+  // The argument we are looking for epilogue ARC instruction for.
+  SILValue Arg;
+
+  /// The allocator we are currently using.
+  llvm::SpecificBumpPtrAllocator<EpilogueARCBlockState> BPA;
+
+  /// Current function we are analyzing.
+  SILFunction *F;
+
+  /// Current post-order we are using.
+  PostOrderFunctionInfo *PO;
+
+  /// Current alias analysis we are using.
+  AliasAnalysis *AA;
+
+  /// Current rc-identity we are using.
+  RCIdentityFunctionInfo *RCFI;
+
+  /// The epilogue retains or releases.
+  llvm::SmallVector<SILInstruction *, 1> EpilogueARCInsts; 
+
+  /// All the retain/release block state for all the basic blocks in the function. 
+  llvm::DenseMap<SILBasicBlock *, EpilogueARCBlockState *> EpilogueARCBlockStates;
```

Comment:
EpilogueARCBlockState is super small. Why not store it as value in the map instead of allocating it with a bumpptrallocator?